lib/ref_arch_setup/download_helper.rb

```diff
@@ -0,0 +1,337 @@
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] version The specified version of PE
+    #
+    # @return [true, false] Whether the specified version is supported
+    #
+    # @example:
+    #   result = ensure_supported_prod_version("2018.1.4")
+    #
+    def self.ensure_supported_prod_version(version)
+      major_version = version.split(".")[0]
+      supported_version = MIN_PROD_VERSION.split(".")[0]
+
+      supported = major_version >= supported_version ? true : false
+      puts "Specified version #{version} is supported by RAS" if supported
+      puts "The minimum supported version is #{MIN_PROD_VERSION}" unless supported
+
+      raise "Specified version #{version} is not supported by RAS" unless supported
+
+      return supported
+    end
+
+    # Fetches the list of production PE versions from the 'Puppet Enterprise Version History'
+    #
+    # @author Bill Claytor
+    #
+    # @return [string] The versions list
+    #
+    # @example:
+    #   versions = fetch_prod_versions
+    #
+    def self.fetch_prod_versions
+      versions = parse_prod_versions_url
+      return versions
+    end
+
+    # Parses the 'Puppet Enterprise Version History'
+    # using the specified xpath and returns the result
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] xpath The xpath to use when parsing the version history
+    #
+    # @return [string] The result
+    #
+    # @example:
+    #   versions = parse_prod_versions_url
+    #   latest_version = parse_prod_versions_url("//table/tbody/tr[1]/td[1]")
+    #   result = parse_prod_versions_url("//table/tbody/tr/td[contains(., '#{version}')]")
+    #
+    def self.parse_prod_versions_url(xpath = "//table/tbody/tr/td[1]")
+      puts "Checking Puppet Enterprise Version History: #{@pe_versions_url}"
+
+      uri = URI.parse(@pe_versions_url)
+      response = Net::HTTP.get_response(uri)
+      raise "Invalid response" unless valid_response?(response)
+
+      document = Oga.parse_html(response.body)
+      result = document.xpath(xpath)
+      return result
+    end
+
+    # Determines whether the response is valid
+    #
+    # @author Bill Claytor
+    #
+    # @param [Net::HTTPResponse] res The HTTP response to evaluate
+    # @param [Array] valid_response_codes The list of valid response codes
+    # @param [Array] invalid_response_bodies The list of invalid response bodies
+    #
+    # @return [true,false] Based on whether the response is valid
+    #
+    # @example
+    #   valid = valid_response?(res)
+    #   valid = valid_response?(res, ["200", "123"], ["", nil], "xyz")
+    #
+    def self.valid_response?(res, valid_response_codes = ["200"],
+                             invalid_response_bodies = ["", nil])
+      # TODO: other criteria?
+      is_valid_response = false
+
+      if res.nil? || !valid_response_codes.include?(res.code) ||
+         invalid_response_bodies.include?(res.body)
+        puts "Invalid response:"
```

Comment:
Generally a "?" method doesn't output text, it just returns true or false and lets the called decide what to do with it.  And since in this case you only call it from one place and raise an error if it fails you have two choices... 1 - move the output text to the caller, probably include it in the raised error, or 2 - change this to validate response and raise the error with all the text from this method.   I like option 2 better in this case, but whatever works for you.

Reply:
Updated with option 2